src/fluree/db/json_ld/shacl.cljc

```diff
@@ -541,6 +561,33 @@
         #?(:clj (Pattern/compile (or valid-flags 0))
            :cljs (js/RegExp. (or valid-flags ""))))))
 
+(defn resolve-path-type
+  "Associate each property path object with its path type in order to govern path flake resolution during validation."
+  [db path-pid]
+  (go-try
+    (if-let [path-flake (->> (<? (query-range/index-range db :spot = [path-pid]))
+                             (remove #(#{const/$xsd:anyURI} (flake/p %)))
+                             (first))]
```

Comment:
This is faster if you add a `:flake-limit 1` to the options map to `index-range` since you throw all but the first flake away anyway. Flake/subject limits and offsets are configured in the [`into-page` helper](https://github.com/fluree/db/blob/cc2d4bd72ddf67410bd2d1fa4515157a08815c42/src/fluree/db/query/range.cljc#L213-L224) when building up the transducers supplied to `tree-chan` for `index-range`.